File: Server/src/main/java/com/github/dr/rwserver/core/Application.java
package com.github.dr.rwserver.core;

import com.github.dr.rwserver.core.thread.Threads;
import com.github.dr.rwserver.data.global.Data;
import com.github.dr.rwserver.data.plugin.PluginData;
import com.github.dr.rwserver.data.plugin.PluginManage;
import com.github.dr.rwserver.net.Administration;
import com.github.dr.rwserver.struct.Seq;
import com.github.dr.rwserver.util.file.FileUtil;
import com.github.dr.rwserver.util.log.Log;

import java.util.UUID;

import static com.github.dr.rwserver.util.IsUtil.isBlank;
import static com.github.dr.rwserver.util.RandomUtil.generateStr;

/**
 * @author Dr
 */
public final class Application {
    public final PluginData pluginData;
    /** 服务器唯一UUID */
    public String serverConnectUuid;
    public String serverToken;
    public Seq<String> unitBase64;
    public Administration admin;
    public boolean upServerList = false;

    public String serverName = "m3ua";
    public float defIncome = 1f;

    public Application() {
        pluginData = new PluginData();
        serverToken = generateStr(40);
    }

    public final void load() {
        pluginData.setFileUtil(FileUtil.getFolder(Data.Plugin_Data_Path).toFile("Settings.bin"));
        admin = new Administration(pluginData);
        serverConnectUuid = pluginData.getData("serverConnectUuid", UUID.randomUUID().toString());
        unitBase64 = pluginData.getData("unitBase64", new Seq<>());

        Threads.addSavePool(() -> {
            pluginData.setData("serverConnectUuid",serverConnectUuid);
            pluginData.setData("unitBase64",unitBase64);
        });
    }

    public void save() {
        // 先执行自己的保存
        Threads.runSavePool();
        // 保存自己
        pluginData.save();
        // 保存Plugin
        PluginManage.runOnDisable();
    }

    public final long getJavaHeap() {
        return Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
    }

    public final long getJavaTotalMemory() {
        return Runtime.getRuntime().totalMemory();
    }

    public final long getJavaFreeMemory() {
        return Runtime.getRuntime().freeMemory();
    }

    public final String getJavaVendor() {
        return get("java.vendor");
    }

    public final String getJavaVersion() {
        return get("java.version");
    }

    public final String getOsName() {
        return get("os.name");
    }

    public final boolean isJavaVersionAtLeast(final float requiredVersion) {
        final String version = get("java.version");
        return (isBlank(version) ? 0f : Float.parseFloat(version)) >= requiredVersion;
    }

    public final boolean isWindows() {
        final String os = get("os.name");
        return (isBlank(os) || os.toLowerCase().contains("windows"));
    }

    public final long getPid() {
        return ProcessHandle.current().pid();
    }


    /**
     * 取得系统属性，如果因为Java安全的限制而失败，则将错误打在Log中，然后返回
     * @param name  属性名
     * @return 属性值或null
     * @see System String
     * @see System String
     */
    private String get(final String name) {
        String value = null;
        try {
            value = System.getProperty(name);
        } catch (SecurityException e) {
            Log.error("Security level limit",e);
        }
        if (null == value) {
            try {
                value = System.getenv(name);
            } catch (SecurityException e) {
                Log.error("Security level limit",e);
            }
        }
        return value;
    }
}

File: Server/src/main/java/com/github/dr/rwserver/core/Application.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.core

import com.github.dr.rwserver.core.thread.Threads.addSavePool
import com.github.dr.rwserver.core.thread.Threads.runSavePool
import com.github.dr.rwserver.data.global.Data
import com.github.dr.rwserver.data.plugin.PluginData
import com.github.dr.rwserver.data.plugin.PluginManage.runOnDisable
import com.github.dr.rwserver.net.Administration
import com.github.dr.rwserver.struct.Seq
import com.github.dr.rwserver.util.IsUtil.isBlank
import com.github.dr.rwserver.util.RandomUtil.generateStr
import com.github.dr.rwserver.util.file.FileUtil
import com.github.dr.rwserver.util.log.Log.error
import java.lang.management.ManagementFactory
import java.util.*


/**
 * @author Dr
 */
class Application {
    private val pluginData: PluginData = PluginData()

    /** 服务器唯一UUID  */
    lateinit var serverConnectUuid: String
    @JvmField
    var serverToken: String = generateStr(40)
    lateinit var unitBase64: Seq<String>
    lateinit var admin: Administration
    @JvmField
    var upServerList = false
    @JvmField
    var serverName = "RW-HPS"

    fun load() {
        pluginData.setFileUtil(FileUtil.getFolder(Data.Plugin_Data_Path).toFile("Settings.bin"))
        admin = Administration(pluginData)

        Initialization.startInit(pluginData)

        serverName = Data.config.ServerName

        serverConnectUuid = pluginData.getData("serverConnectUuid") { UUID.randomUUID().toString() }
        unitBase64 = pluginData.getData("unitBase64") {Seq()}
        addSavePool {
            pluginData.setData("serverConnectUuid", serverConnectUuid)
            pluginData.setData("unitBase64", unitBase64)
        }
    }

    fun save() {
        // 先执行自己的保存
        runSavePool()
        // 保存自己
        pluginData.save()
        // 保存Plugin
        runOnDisable()
    }

    val javaHeap: Long
        get() = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
    val javaTotalMemory: Long
        get() = Runtime.getRuntime().totalMemory()
    val javaFreeMemory: Long
        get() = Runtime.getRuntime().freeMemory()
    val javaVendor: String
        get() = get("java.vendor")
    val javaVersion: String
        get() = get("java.version")
    val osName: String
        get() = get("os.name")

    fun isJavaVersionAtLeast(requiredVersion: Float): Boolean {
        val version = get("java.version")
        return (if (isBlank(version)) 0f else version.toFloat()) >= requiredVersion
    }

    val isWindows: Boolean
        get() {
            val os = get("os.name")
            return isBlank(os) || os.lowercase(Locale.getDefault()).contains("windows")
        }
    val pid: Long
        // 唯一的到Java11理由
        // get() = ProcessHandle.current().pid()
        // 现在理由它没了
        get() = jvmPid()


    private fun jvmPid(): Long {
        return try {
            var pid = ManagementFactory.getRuntimeMXBean().name
            val indexOf = pid.indexOf('@')
            if (indexOf > 0) {
                pid = pid.substring(0, indexOf)
            }
            pid.toLong()

        } catch (e: Exception) {
            -1
        }
    }

    /**
     * 取得系统属性，如果因为Java安全的限制而失败，则将错误打在Log中，然后返回
     * @param name  属性名
     * @return 属性值或null
     * @see System String
     *
     * @see System String
     */
    private operator fun get(name: String): String {
        var value: String? = null
        try {
            value = System.getProperty(name)
        } catch (e: SecurityException) {
            error("Security level limit", e)
        }
        if (null == value) {
            try {
                value = System.getenv(name)
            } catch (e: SecurityException) {
                error("Security level limit", e)
            }
        }
        return value!!
    }

}